### ISSUE TITLE
Token validation and error 411: LengthRequired

### ISSUE THREAD

**Issue by SubZero0** [RogueException/Discord.Net#1577]
For those that are lazy, there's a tl;dr.

## Issue

While reading the Discord.Net channel in Discord API, I saw someone asking why their `AddRoleAsync` was throwing a `The server responded with error 411: LengthRequired`, later they found out the reason: there's a new line in the file that they stored their token.

I thought it was odd they actually were able to log in on Rest and Gateway without issues and tried to test it myself and tested that Strings that end, start, or have `\r`, `\n`, or a space would throw an exception when validating but Discord.Net won't stop the program flow. That would be followed by Discord failing the Gateway authorization.

But there was a case where it wouldn't. If the token ended with `\r\n`, it would sucessfully connect to the Gateway as well, but when doing a POST request, it would throw that error (probably because the line break malformed the request).

So while they indeed missed the first warning from the token validation, I think it would be optimal to either fix this inside Discord.Net or add it somewhere in the documentation (example: common case for an error 411).

## Tl;dr
Tokens that end with `\r\n` will succesfully log in to the Gateway and do GET requests, but throw an error 411: LengthRequired when doing other http methods like POST. This isn't expected or intuitive since it would be expected a failed authorization for both.

## Fix

So I thought about a few ways to fix it:
1. Block the program flow when the token validation fails with an invalid character;
2. Trim the token (basically silently fix it);
3. Add it somewhere to the documentation since the error isn't really intuitive.

Any ideas or other suggestions?

**Comment by Hona** [comment]
Can confirm this issue was about my bot, and the newline was the issue

**Comment by Emzi0767** [comment]
\r\n is a canonical end of line in HTTP spec. This could mean that a header with this at the end causes a double line break, thus ending the header section of the request, per spec. Someone should dump traffic to see what the deal is, but it may be a bug in System.Net.Http, or discord decoding something too eagerly on their end.

Either way the solution is to trim tokens passed to config.

**Comment by SubZero0** [comment]
I ended reading a bit more about it and I would expect a FormatException from the [HttpHeaders.Add(String, String)](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpheaders.add?view=netcore-3.1#System_Net_Http_Headers_HttpHeaders_Add_System_String_System_String_) that never happens, so guess we will need to work around it.

Not sure why the decision to not stop the flow if the token is invalid, but I believe a nonintuitive error long after the bot successfully logged in is even worse, so with the current flow in mind (the validation could be turned off as well) and Emzi's comment, I decided to really  go with the 2nd.
